mlir/tools/rocmlir-gen/rocmlir-gen.cpp

```diff
@@ -3373,6 +3379,45 @@ static ModuleOp generateKernel(MLIRContext *context, GenParams &genParams,
   return module;
 }
 
+static ModuleOp populateCloneHarnessLogic(ModuleOp &module)
+{
+  MLIRContext *context = module.getContext();
+  OpBuilder b(context);
+  func::FuncOp originalFunc = module.lookupSymbol<func::FuncOp>(testFuncName);
+  auto readAttr =
+      b.getNamedAttr(func::FuncOp::getReadAccessAttrName(), b.getUnitAttr());
+  auto writeAttr =
+      b.getNamedAttr(func::FuncOp::getWriteAccessAttrName(), b.getUnitAttr());
+  for (size_t index = 0; index < originalFunc.getArguments().size(); index++)
+    originalFunc.setArgAttrs(index, readAttr);
+  for (size_t index = 0; index < originalFunc.getNumResults(); index++)
+    originalFunc.setResultAttrs(index, writeAttr);
+  auto loc = originalFunc->getLoc();
+  auto wrapperFunc = func::FuncOp::create(loc, testFuncName + "_wrapper",
+                                          originalFunc.getFunctionType());
+  Block *block = wrapperFunc.addEntryBlock();
+  b.setInsertionPointToStart(block);
+  auto launchOp = b.create<mhal::LaunchOp>(loc, originalFunc, ValueRange{},
+                                           block->getArguments());
+  auto results = launchOp->getResults();
+  b.create<mhal::AwaitOp>(loc, results.front());
+  b.create<func::ReturnOp>(loc, ValueRange{results.drop_front()});
+  module.push_back(wrapperFunc);
+
+  auto xmoduleOp = ModuleOp::create(b.getUnknownLoc(), "__xmodule_");
+  StringAttr archAttr = b.getStringAttr(arch);
+  xmoduleOp->setAttr("mhal.arch", archAttr);
+  xmoduleOp->setAttr("mhal.module", b.getUnitAttr());
+  auto *cloneFunc = originalFunc->clone();
+  auto cloneFuncOp = dyn_cast<func::FuncOp>(cloneFunc);
+  cloneFuncOp->setAttr("kernel", b.getUnitAttr());
+  cloneFuncOp->setAttr("original_func", SymbolRefAttr::get(originalFunc));
+  xmoduleOp.push_back(cloneFuncOp);
+  module.push_back(xmoduleOp);
+  //module.dump();
+  return module;
+}
```

Comment:
Note: please remove the `kernel` attribute from the original function if it's present